Equinix metal provider

mdemeke opened the pull request:
-Added Equinix as a Provider
-Updated README
-Tested using c3.small.x86 machine
-To make benchmark diagnostics simple and less time consuming
-Equinix CLI required for use (Reference Read me for instructions)

mdemeke committed:
Added Equinix Metal Provider: Integration configured and  modified ReadMe

mdemeke committed:
UPDATED

mdemeke committed:
UPDATED

mdemeke committed:
Update __init__.py

jeremytanner commented on README.md:
````diff
@@ -261,7 +261,11 @@ $ ./pkb.py --cloud=Rackspace --machine_type=general1-2 --benchmarks=iperf
 ```bash
 $ ./pkb.py --cloud=ProfitBricks --machine_type=Small --benchmarks=iperf
 ```
+## Example run on Metal
````
Equinix Metal

jeremytanner commented on README.md:
````diff
@@ -261,7 +261,11 @@ $ ./pkb.py --cloud=Rackspace --machine_type=general1-2 --benchmarks=iperf
+## Example run on Metal
 
+```bash
+./pkb.py --cloud=Equinix --machine_type=c3.small.x86 --image=ubuntu_18_04 --benchmarks=iperf 
````
EquinixMetal

jeremytanner commented on README.md:
```diff
@@ -349,6 +353,7 @@ CloudStack | QC-1 | |
 Rackspace | IAD | OnMetal machine-types are available only in IAD zone
 Kubernetes | k8s | |
 ProfitBricks | AUTO | Additional zones: ZONE_1, ZONE_2, or ZONE_3
+EquinixMetal | da | A metro is a regional location where in this case it is Dallas.
```
Zone is mapped to metro. Equinix Metal Metros are facilities that are grouped together geographically and share capacity and networking features.

jeremytanner commented on perfkitbenchmarker/providers/README.md:
```diff
@@ -462,3 +462,7 @@ Open the `.boto` file and edit the following fields:
+    
+## Equinix Metal Config
+Install `metal`, the Equinix CLI, following the instructions at
+    `https://github.com/equinix/metal-cli`.
```
Change to clickable link, add init instructions similar to DigitalOcean

jeremytanner commented on perfkitbenchmarker/providers/__init__.py:
```diff
@@ -33,6 +32,7 @@
 ALICLOUD = 'AliCloud'
 KUBERNETES = 'Kubernetes'
 DIGITALOCEAN = 'DigitalOcean'
+EQUINIX = 'Equinix'
```
EQUINIXMETAL & EquinixMetal

jeremytanner commented on perfkitbenchmarker/providers/__init__.py:
```diff
@@ -43,7 +43,7 @@
 
 VALID_CLOUDS = (GCP, AZURE, AWS, IBMCLOUD, DIGITALOCEAN, KUBERNETES, OPENSTACK,
-                RACKSPACE, CLOUDSTACK, ALICLOUD, MESOS, PROFITBRICKS, DOCKER)
+                RACKSPACE, CLOUDSTACK, ALICLOUD, MESOS, PROFITBRICKS, DOCKER, EQUINIX)
```
EQUINIXMETAL

jeremytanner commented on perfkitbenchmarker/providers/equinix/equinix_virtual_machine.py:
```diff
@@ -0,0 +1,117 @@
+    """Returns true if the VM exists."""
+
+    response, retcode = util.MetalAndParse(
+        ['device', 'get', self.device_id])
```
Without the -i flag, this will return all devices in the project instead of the specified device

jeremytanner commented on perfkitbenchmarker/providers/equinix/provider_info.py:
```diff
@@ -0,0 +1,27 @@
+class EquinixProviderInfo(provider_info.BaseProviderInfo):
+
+ 
+  CLOUD = providers.EQUINIX
```
EQUINIXMETAL

jeremytanner commented on perfkitbenchmarker/providers/equinix/util.py:
```diff
@@ -0,0 +1,99 @@
+      - Device's return code
+
+  Raises:
+    errors.VmUtil.CalledProcessError if doctl fails.
```
This comment refers to doctl, not used in this file

jeremytanner commented on perfkitbenchmarker/providers/equinix/equinix_virtual_machine.py:
```diff
@@ -0,0 +1,117 @@
+    for interface in response['ip_addresses']:
+      if interface['address_family'] != 4:
+        continue
+      if interface['public'] == True:
```
`if interface['public']:` should be sufficient, though this isn't always parsing a public IP for the created machine. (Test against the `sv` metro)

jeremytanner commented on perfkitbenchmarker/providers/README.md:
```diff
@@ -9,7 +9,7 @@ provider(s) you want to benchmark:
+*   [Equinix Metal](https://metal.equinix.com)
 ## Cloud account setup
 
 This section describes the setup steps needed for each cloud system. Note that
```
This section is missing a link to the Equinix Metal cloud setup steps

jeremytanner commented on perfkitbenchmarker/providers/equinix/equinix_virtual_machine.py:
```diff
@@ -0,0 +1,117 @@
+        ['device', 'get', self.device_id])
+
+    return retcode == 0
+#Disk creation needs to be Finished, therefore FIO testing errors will occur unless changed
```
Remove TODOs

jeremytanner reviewed (changes requested):
Tested `./pkb.py --cloud=Equinix --machine_type=c3.small.x86 --zone=sv --image=ubuntu_18_04 --benchmarks=coremark` keeps looping as the machine is created, but the IP isn't picked up: VM: None / ssh -A -p 22 perfkit@None

Tried the same in da zone / metro, while an IP is picked up, errors occur on machine cleanup: Exception occurred while calling DeleteVm

JThomasEquinix commented:
Looks great to me everything checks out

mdemeke committed:
Update README.md

mdemeke committed:
Update README.md

mdemeke committed:
Create README.md

mdemeke committed:
Update README.md

mdemeke committed:
Update README.md

mdemeke committed:
Update equinix_virtual_machine.py

mdemeke committed:
Update __init__.py

mdemeke committed:
Update equinix_virtual_machine.py

mdemeke committed:
Update equinix_virtual_machine.py

mdemeke committed:
Update provider_info.py

mdemeke committed:
Update util.py

mdemeke committed:
Update default_config_constants.yaml

mdemeke committed:
Update util.py

mdemeke reviewed (commented):
Improvements and adjustments made

mdemeke commented:
-Added Equinix as a Provider
-Updated README
-Tested using c3.small.x86 machine
-To make benchmark diagnostics simple and less time consuming
-Equinix CLI required for use (Reference Read me for instructions)
